src/main/scala/components/memory/memory-combin-ports.scala

```diff
@@ -0,0 +1,131 @@
+// Combinational/'synchronous' memory module
+
+package dinocpu
+
+import chisel3._
+import chisel3.util._
+import dinocpu.MemoryOperation._
+
+/**
+  * The instruction memory port.
+  *
+  * The I/O for this module is defined in [[IMemPortIO]].
+  */
+class ICombinMemPort extends BaseIMemPort {
+  // When the pipeline is supplying a high valid signal
+  when (io.valid) {
+    val request = Wire(new Request)
+    request := DontCare
+    request.address   := io.address
+    request.operation := Read
+
+    io.request.bits  := request
+    io.request.valid := true.B
+  } .otherwise {
+    io.request.valid := false.B
+  }
+
+  // When the memory is outputting a valid instruction
+  io.good := io.response.valid
+  when (io.response.valid) {
+    io.instruction := io.response.bits.data
+  }
+}
+
+/**
+  * The data memory port.
+  *
+  * The I/O for this module is defined in [[DMemPortIO]].
+  */
+class DCombinMemPort extends BaseDMemPort {
+  io.good := io.response.valid && io.memread && !io.memwrite
+
+  when (io.valid && (io.memread || io.memwrite)) {
+    // Check that we are not issuing a read and write at the same time
+    assert(!(io.memread && io.memwrite))
+
+    io.request.bits.address := io.address
+    io.request.valid := true.B
+
+    when (io.memwrite) {
+      // We issue a ReadWrite to the backing memory.
+      // Basic run-down of the ReadWrite operation:
+      // - DCombinMemPort sends a ReadWrite at a specific address, **addr**.
+      // - Backing memory outputs the data at **addr** in io.response
+      // - DCombinMemPort notes that io.memwrite is high in the response path. io.response.bits.data
+      //   is masked and sign extended, and sent down io.request.writedata
+      // - Backing memory receives the modified writedata and feeds it into the memory at **addr**.
+      // Since this is combinational logic, this should theoretically all resolve in one clock cycle with no issues
+      io.request.bits.operation := ReadWrite
+    } .otherwise {
+      // Issue a normal read to the backing memory
+      io.request.bits.operation := Read
+    }
+  } .otherwise {
+    // no request coming in so don't send a request out
+    io.request.valid := false.B
+  }
+
+  // Response path
+  when (io.response.valid) {
+    when (io.memwrite) {
+      // Perform writedata modification and send it down io.request.writedata.
+      val writedata = Wire (UInt (32.W))
+
+      // When not writing a whole word
+      when (io.maskmode =/= 2.U) {
+        // Read in the existing piece of data at the address, so we "overwrite" only part of it
+        val offset = io.address (1, 0)
+        val readdata = Wire (UInt (32.W))
+
+        readdata := io.response.bits.data
+
+        val data = Wire (UInt (32.W))
+        // Mask the portion of the existing data so it can be or'd with the writedata
+        when (io.maskmode === 0.U) {
```

Comment:
I assume the plan is to split this logic out into a function in the `BaseDMemPort` in another patch since the code is replicated in `NonCombinMemPort`.